presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -2116,9 +2166,22 @@ private static void renameDirectory(HdfsContext context, HdfsEnvironment hdfsEnv
         runWhenPathDoesntExist.run();
 
         try {
-            if (!hdfsEnvironment.getFileSystem(context, source).rename(source, target)) {
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(context, source);
+            if (!fileSystem.rename(source, target)) {
                 throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Failed to rename %s to %s: rename returned false", source, target));
             }
+            // setOwner failure shouldn't cause sql failure
+            try {
+                if (warehouseSubdirInheritPerms) {
+                    FileStatus fileStatus = fileSystem.getFileStatus(target.getParent());
+                    String owner = fileStatus.getOwner();
+                    String group = fileStatus.getGroup();
+                    recursiveSetOwner(fileSystem, target, owner, group, true);
```

Comment:
We could avoid recursive set owner, if we made sure that files are created in tmp with appropriate permissions.
We could apply the permissions (based on the target path, not the parent) right after creating the file in tmp,
on a worker, rather than doing it here recursively.